V1 - Vertex AI: Qwikstart lab for SPL

dougkelly opened the pull request:
* Train a TensorFlow model locally in a hosted Vertex Notebook.
* Create a managed Tabular dataset artifact for experiment tracking.
* Containerize your training code with Cloud Build and push it to Google Cloud Artifact Registry.
* Run a Vertex AI custom training job with your custom model container.
* Use Vertex TensorBoard to visualize model performance.
* Deploy your trained model to a Vertex Prediction Endpoint for serving predictions.
* Request an online prediction and explanation and see the response.

dougkelly commented:
@danielsparing I would greatly appreciate any and all feedback you have time to give by EOD tomorrow. Apologies for the delay from when we chatted MOnday and short notice, spend significant time wrangling with TB service.

@sanjanalreddy keen on your feedback as well if you have time tomorrow :)

- lab design and narrative (this isn't a challenge lab so it's more of a walkthrough but any ideas on meaningful TODOs?)
- anything not work? (TB uploader was super flaky for me, curious if either of you experience similar and have ideas on improving it; also is that section needed?)
- Any suggestions to improve code readability, succinctness?

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -1213,68 +1349,74 @@
    "source": [
-    "print(model.evaluate(evalds))"
+    "history = model.fit(trainds,\n",
+    "                    validation_data=evalds,\n",
```
you earlier renamed `evalds` to `devds`, so here evalds not found

dougkelly replied:
Done

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/online-retail-clv-3M/trainer/task.py:
```diff
@@ -0,0 +1,34 @@
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    # Vertex custom container training args. These are set by Vertex AI during training can be overwritten.
```
you corrected this sentence in the %%writefile as following, so we can update here too:
    # Vertex custom container training args. These are set by Vertex AI during training but can also be overwritten.

dougkelly replied:
Thanks, done.

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -1567,158 +1717,384 @@
+    "%%writefile {MODEL_NAME}/Dockerfile\n",
+    "# Specifies base image and tag.\n",
+    "# https://cloud.google.com/vertex-ai/docs/training/pre-built-containers\n",
+    "# us-docker.pkg.dev/vertex-ai/training/tf-cpu.2-3:latest\n",
```
See my comment above, I believe pre-built containers are not relevant here -- or we should go all in and use pre-built containers, with no Dockerfile

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -1567,158 +1717,384 @@
+   ],
+   "source": [
+    "%%writefile {MODEL_NAME}/requirements.txt\n",
+    "explainable-ai-sdk==1.3.0\n",
```
comment: `explainable-ai-sdk` is included out of the box in the pre-built containers i believe, but it still makes sense to leave this in, _if_ we anticipate breaking changes in a version upgrade

dougkelly replied:
https://cloud.google.com/vertex-ai/docs/training/pre-built-containers

I didn't see it on here actually which was surprising. The other reason for this is the tensorflow IO package (not part of core) for the BigQuery reader. Pyarrow is a dependency for that.

danielsparing replied:
Oh I am probably wrong, sorry! It might be in the Notebooks but not in the actual pre-built containers. So leave it in, if in doubt!

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -1567,158 +1717,384 @@
    "metadata": {},
    "source": [
-    "Next, you will use [Cloud Build](https://cloud.google.com/build) to build and upload your custom TensorFlow model container to [Google Cloud Container Registry](https://cloud.google.com/container-registry).\n",
+    "These are additional dependencies for your model code outside the pre-built containers needed for prediction explainability and the BigQuery TensorFlow IO reader."
```
It is possible to supply Python dependencies to pre-built containers too.

So either we need to move to true pre-built containers, or identify another reason why we need custom.

https://cloud.google.com/vertex-ai/docs/training/create-python-pre-built-container#working_with_dependencies

dougkelly replied:
Updated explanation. Packaging ML model code + dependencies in single container brings portability, stronger reproducibility guarantees, and onboarding workflow to Vertex Pipelines.

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -1734,127 +2110,188 @@
-    "    container_uri=f\"gcr.io/{PROJECT_ID}/online-retail-clv:latest\",\n",
+    "    display_name=\"online-retail-clv-3M-dnn-regressor\",\n",
+    "    container_uri=IMAGE_URI,\n",
     "    # https://cloud.google.com/vertex-ai/docs/predictions/pre-built-containers\n",
```
Either we should use true pre-built containers, from us-docker.pkg.dev/vertex-ai/training and then use CustomTrainingJob or CustomPythonPackageTrainingJob, or pre-built containers are not relevant here

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -1567,158 +1717,384 @@
-    "    save_weights_only=True,\n",
-    "    monitor='val_loss',\n",
-    "    mode='min')\n",
+    "Notice the base image below is a Vertex AI [pre-built container](https://cloud.google.com/vertex-ai/docs/training/pre-built-containers). This image includes dependencies you need for your training code such TensorFlow and BigQuery SDK which are already tested and built to save you time when building your custom container."
```
The one below isn't a Vertex pre-built container, it is a Deep Learning Container.

I am not sure if Vertex pre-built containers are even supposed to work with custom container methods.

dougkelly replied:
Thanks, updating. Sloppiness/sleepiness on my part, will update to say its using DL container.

danielsparing replied:
Oh I know they are confusing -- and I am sure that the Vertex containers are closely built on the DLC but not the same.

Similar to how I learned that a Vertex Notebook based on a TF2.5 DLC is very similar to, but not at all identical with a TF2.5 DLVM.

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -2065,25 +2501,31 @@
   {
    "cell_type": "markdown",
-   "id": "b16d5016",
+   "id": "90370413",
```
I suggest to include a lightweight cleanup section with `endpoint.delete()` , because otherwise an endpoint occurs ongoing charges (as its minimum min_replica_count is 1)

danielsparing reviewed (commented):
My main comment is that I do not think it is supported to use a Vertex AI pre-built container as a basis of a custom container training (and you are currently not using it either), so we should either
- try to make the pre-built container work (lightweight pip dependencies are possible), or
- make it clear that we are fully using custom container, based on a Deep Learning Container.

There are other dependency and variable changes that need to be rewired for it to work end-to-end.

In the interest of time I am submitting this before I could have tested beyond the container build part, so you can have a look.

sanjanalreddy commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/online-retail-clv-3M/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+explainable-ai-sdk==1.3.0
+tensorflow-io==0.16.0
+pyarrow
```
Add these to the requirements file
wget
xlrd
openpyxl

dougkelly replied:
Thanks, adding these to the lab requirements.txt outside the container as these should just be dependencies for the data download script.

sanjanalreddy reviewed (commented):
Added a few comments on things which didn't work in ReviewNB when I ran the notebook. 

Changing the table name to the right project ID in the BQ sql statements is a bit frustrating since there are a few queries but I understand that there isn't a better way to do it right now.

sanjanalreddy commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -2,60 +2,72 @@
  "cells": [
```
<p>Just realized you can actually drop the PROJECT ID from and just do</p><pre class="ql-syntax" spellcheck="false">FROM
  online_retail.online_retail_clv_ml
</pre><p>This way the person working on the notebook can actually just run the queries without replacing each FROM statement</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1570/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='26'/>

dougkelly replied:
Cash $$$! Thank you! Fixed.

danielsparing commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -2,60 +2,72 @@
  "cells": [
```
<pre class="ql-syntax" spellcheck="false">Line #1.    !tb-gcp-uploader --tensorboard_resource_name=$TB_INSTANCE_NAME \
</pre><p>This fails for me with the below error message. Probably we need to set up service account permissions, see <a href="https://cloud.google.com/vertex-ai/docs/experiments/tensorboard-training#create_a_service_account_with_required_permissions" target="_blank">here</a> </p><p> </p><pre class="ql-syntax" spellcheck="false">google.api_core.exceptions.PermissionDenied: 403 Permission 'aiplatform.tensorboards.get' denied on resource '//aiplatform.googleapis.com/projects/[...]/locations/us-central1/tensorboards/[...]' (or it may not exist).
</pre><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1570/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='106'/>

dougkelly replied:
Thanks, looking to replicate this on refresh instance. Sanjana and myself didn't encounter this. Maybe one of the APIs in the Readme.

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/README.md:
```diff
@@ -1,49 +1,49 @@
+* Containerize your training code with [**Cloud Build**](https://cloud.google.com/build) and push it to [**Google Cloud Artifact Registry**](https://cloud.google.com/artifact-registry).
+* Run a [**Vertex AI custom training job**](https://cloud.google.com/vertex-ai/docs/training/custom-training) with your custom model container.
 * Use [**Vertex TensorBoard**](https://cloud.google.com/vertex-ai/docs/experiments/tensorboard-overview) to visualize model performance.
 * Deploy your trained model to a [**Vertex Prediction Endpoint**](https://cloud.google.com/vertex-ai/docs/predictions/getting-predictions) for serving predictions.
```
Thanks, SGTM. Updating.

danielsparing commented:
@dougkelly I created a version with pre-built container training, you could best view it via the ReviewNB link in the PR.
https://github.com/GoogleCloudPlatform/training-data-analyst/pull/1571

I believe this way the flow would be much faster as you don't have to wait 15+ minutes for the container build.

Feel free to ignore the actual PR and do your changes as you prefer

sanjanalreddy commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -2,60 +2,72 @@
  "cells": [
```
<pre class="ql-syntax" spellcheck="false">Line #2.                        validation_data=evalds,
</pre><p>evalds here should be devds</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/training-data-analyst/pull/1570/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='61'/>

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/README.md:
```diff
@@ -1,49 +1,49 @@
 # Vertex AI: Qwikstart
 
-In this lab, you will walk through an introductory, end-to-end training and deployment workflow for a custom container TensorFlow model on Vertex AI.
+In this lab, you will use [BigQuery](https://cloud.google.com/bigquery) for data processing and exploratory data analysis and the [Vertex AI](https://cloud.google.com/vertex-ai) platform to train and deploy a custom TensorFlow Regressor model to predict customer lifetime value. The goal of the lab is to introduce to Vertex AI through a high value real world use case - predictive CLV. You will sta ...
```
Done. In the header and included abbreviation in text again in first line.

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/requirements.txt:
```diff
@@ -0,0 +1,2 @@
+google-cloud-aiplatform
```
Done

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -167,97 +170,188 @@
+    "BQ_RAW_TABLE_NAME=\"online_retail_raw\"\n",
+    "BQ_CLEAN_TABLE_NAME=\"online_retail_clean\"\n",
+    "BQ_ML_TABLE_NAME=\"online_retail_ml\"\n",
+    "BQ_URI=f\"bq://{PROJECT_ID}.{BQ_DATASET_NAME}.{BQ_TABLE_NAME}\""
```
Thanks yes, fixed.

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -167,97 +170,188 @@
    "outputs": [],
-   "source": []
+   "source": [
+    "!python utils/data_download.py \\\n",
```
Done thanks.

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -167,97 +170,188 @@
-    "FROM\n",
-    "`dougkelly-vertex-demos.online_retail.online_retail_clv_ml`"
+    "FROM \n",
+    "  `dougkelly-vertex-demos.online_retail.online_retail_clv_ml` #Please update this with your table name e.g. `{PROJECT_ID}.{BQ_DATASET_NAME}.{BQ_TABLE_NAME}`."
```
Removed PROJECT ID, works great

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -167,97 +170,188 @@
+    "BQ_DATASET_NAME=\"online_retail\"\n",
+    "BQ_RAW_TABLE_NAME=\"online_retail_raw\"\n",
+    "BQ_CLEAN_TABLE_NAME=\"online_retail_clean\"\n",
+    "BQ_ML_TABLE_NAME=\"online_retail_ml\"\n",
```
Thanks fixed.

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/README.md:
````diff
@@ -1,49 +1,49 @@
-3. Use `gcloud` to enable the services
+1.3. Use `gcloud` to enable the services
 ```
 gcloud services enable \
````
Thanks done.

dougkelly commented on self-paced-labs/vertex-ai/vertex-ai-qwikstart/lab_exercise.ipynb:
```diff
@@ -167,97 +170,188 @@
-   "source": []
+   "source": [
+    "!python utils/data_download.py \\\n",
+    "--PROJECT_ID={PROJECT_ID} \\\n",
```
Done.

dougkelly commented:
Thank you both! Addressed feedback and clarified the custom container workflow of building on the DLVM images as packaging model code + dependencies into single container as bringing additional portability and reproducibility guarantees, as well as onramp to ML Pipelines workflow. There will be other opportunities to highlight the Python script or Python Package + pre-built workflow which is the faster option. Going ahead and merging due to time constraints.
